# Comparing runtime to evaluate FOM and its gradient for different figures of merit

In [3]:
"""
Test the runtime of figure of mert functions.
"""

import numpy as np

import sys
sys.path.append("../")
sys.path.append("../Optimisation")

import time

import Optimisation.opt as opt
from parameters import Parameters, D1_ND
I0, L, m, c = Parameters()
from twobox import TwoBox

Nx = 100
nG = 25
Doppler = D1_ND([0.2,0])


# Test gratings
standard_params = [1.4, 0.7, 0.15, 0.35, 0.6, 9., 9., 20., 1., 1.45**2]
standard_grating = TwoBox(*standard_params, wavelength=1., angle=0., Nx=Nx, nG=nG, Qabs=np.inf)
# standard_grating.show_permittivity(show_analytic_box=True)

near_cutoff_params = [1.999, 0.7, 0.15, 0.35, 0.6, 9., 9., 20., 1., 1.45**2] 
near_cutoff_grating = TwoBox(*near_cutoff_params, wavelength=1., angle=0., Nx=Nx, nG=nG, Qabs=np.inf)

thin_pillars_params = [1.4, 0.7, 0.01, 0.01, 0.6, 9., 9., 20., 1., 1.45**2]
thin_pillars_grating = TwoBox(*thin_pillars_params, wavelength=1., angle=0., Nx=Nx, nG=nG, Qabs=np.inf)

closebox_params = [1.4, 0.7, 0.15, 0.35, 0.27, 9., 9., 20., 1., 1.45**2]
closebox_grating = TwoBox(*closebox_params, wavelength=1., angle=0., Nx=Nx, nG=nG, Qabs=np.inf)

overlap_params = [1.4, 0.7, 0.15, 0.35, 0.15, 12., 9., 20., 1., 1.45**2]
overlap_grating = TwoBox(*overlap_params, wavelength=1., angle=0., Nx=Nx, nG=nG, Qabs=np.inf)

clipbox_params = [1.4, 0.7, 0.35, 0.9, 1.1, 9., 9., 20., 1., 1.45**2]
clipbox_grating = TwoBox(*clipbox_params, wavelength=1., angle=0., Nx=Nx, nG=nG, Qabs=np.inf)


def FOM(grating: TwoBox, goal: float=0.1, return_grad: bool=True):
    return opt.FOM_uniform(grating, 5., goal, return_grad)

goal = 0.1
n = 1
grating = near_cutoff_grating

In [4]:
t0 = time.time()
for i in range(n): 
    FOM(grating, goal, False)
t1 = time.time()

total_n = t1-t0
time_per_run = total_n/n
print(time_per_run)

1.346618890762329


In [5]:
t0 = time.time()
for i in range(n): 
    opt.FOM_uniform(grating, goal, 0.1, True)
t1 = time.time()

total_n = t1-t0
time_per_run = total_n/n
print(time_per_run)

8.948760032653809


## Standard grating

### MdS FOM:

Without gradient (s): 1.2317434549331665

With gradient (s): 10.956045866012573


### LvR FOM: 

Without gradient (s): 1.4968903064727783

With gradient (s): 12.115294933319092


## Thin pillars grating

### MdS FOM:

Without gradient (s): 0.8706028461456299

With gradient (s): 7.278752088546753


### LvR FOM: 

Without gradient (s): 1.1792478561401367

With gradient (s): 6.930310010910034


## Near cutoff grating

### MdS FOM:

Without gradient (s): 1.440718173980713

With gradient (s): 7.278752088546753


### LvR FOM: 

Without gradient (s): 1.346618890762329

With gradient (s): 8.948760032653809